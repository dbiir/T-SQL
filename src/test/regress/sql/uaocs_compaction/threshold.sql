-- @Description Tests the basic behavior of (lazy) vacuum w.r.t. to the threshold guc.
CREATE TABLE uaocs_threshold (a INT, b INT, c CHAR(128)) WITH (appendonly=true, orientation=column) distributed by (b);
CREATE INDEX uaocs_threshold_index ON uaocs_threshold(a);
INSERT INTO uaocs_threshold SELECT i as a, 1 as b, 'hello world' as c FROM generate_series(1, 100) AS i;

\set QUIET off
VACUUM uaocs_threshold;
DELETE FROM uaocs_threshold WHERE a < 4;
SELECT COUNT(*) FROM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
-- 97 visible tuples, no vacuum
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
DELETE FROM uaocs_threshold WHERE a < 12;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
-- 89 visible tuples, do vacuum
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
-- no invisible tuples, no vacuum
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
DELETE FROM uaocs_threshold WHERE a < 15;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
-- 3 invisible tuples, no vacuum
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
SET gp_appendonly_compaction_threshold=2;
-- 3 invisible tuples, no vacuum
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');

INSERT INTO uaocs_threshold SELECT i as a, i as b, 'hello world' as c FROM generate_series(100, 200) AS i;
DELETE FROM uaocs_threshold WHERE a > 100 and a < 175;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
VACUUM uaocs_threshold;
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');

-- The percentage of hidden tuples should be 10.1%
-- The threshold guc is set to 10%
SET gp_appendonly_compaction_threshold=10;
CREATE TABLE uaocs_threshold_boundary(a int, b int) WITH(appendonly=TRUE, orientation=COLUMN) distributed by(a);
INSERT INTO uaocs_threshold_boundary SELECT 1, i from generate_series(1, 1000) i;
DELETE FROM uaocs_threshold_boundary WHERE b < 102;
VACUUM uaocs_threshold_boundary;
SELECT * FROM gp_toolkit.__gp_aovisimap_compaction_info('uaocs_threshold_boundary'::regclass);
