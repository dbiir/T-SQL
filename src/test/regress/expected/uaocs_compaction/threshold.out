-- @Description Tests the basic behavior of (lazy) vacuum w.r.t. to the threshold guc.
CREATE TABLE uaocs_threshold (a INT, b INT, c CHAR(128)) WITH (appendonly=true, orientation=column) distributed by (b);
CREATE INDEX uaocs_threshold_index ON uaocs_threshold(a);
INSERT INTO uaocs_threshold SELECT i as a, 1 as b, 'hello world' as c FROM generate_series(1, 100) AS i;
\set QUIET off
VACUUM uaocs_threshold;
VACUUM
DELETE FROM uaocs_threshold WHERE a < 4;
DELETE 3
SELECT COUNT(*) FROM uaocs_threshold;
 count 
-------
    97
(1 row)

SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     1 |      100 |     1
(1 row)

-- 97 visible tuples, no vacuum
VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     1 |      100 |     1
(1 row)

DELETE FROM uaocs_threshold WHERE a < 12;
DELETE 8
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     1 |      100 |     1
(1 row)

-- 89 visible tuples, do vacuum
VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     2 |       89 |     1
     1 |        0 |     1
(2 rows)

-- no invisible tuples, no vacuum
VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     2 |       89 |     1
     1 |        0 |     1
(2 rows)

DELETE FROM uaocs_threshold WHERE a < 15;
DELETE 3
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     2 |       89 |     1
     1 |        0 |     1
(2 rows)

-- 3 invisible tuples, no vacuum
VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     2 |       89 |     1
     1 |        0 |     1
(2 rows)

SET gp_appendonly_compaction_threshold=2;
SET
-- 3 invisible tuples, no vacuum
VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     1 |       86 |     1
     2 |        0 |     1
(2 rows)

INSERT INTO uaocs_threshold SELECT i as a, i as b, 'hello world' as c FROM generate_series(100, 200) AS i;
INSERT 0 101
DELETE FROM uaocs_threshold WHERE a > 100 and a < 175;
DELETE 74
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     2 |        0 |     1
     1 |      187 |     1
(2 rows)

VACUUM uaocs_threshold;
VACUUM
SELECT DISTINCT segno, tupcount, state FROM gp_toolkit.__gp_aocsseg('uaocs_threshold');
 segno | tupcount | state 
-------+----------+-------
     1 |        0 |     1
     2 |      113 |     1
(2 rows)

-- The percentage of hidden tuples should be 10.1%
-- The threshold guc is set to 10%
SET gp_appendonly_compaction_threshold=10;
SET
CREATE TABLE uaocs_threshold_boundary(a int, b int) WITH(appendonly=TRUE, orientation=COLUMN) distributed by(a);
CREATE TABLE
INSERT INTO uaocs_threshold_boundary SELECT 1, i from generate_series(1, 1000) i;
INSERT 0 1000
DELETE FROM uaocs_threshold_boundary WHERE b < 102;
DELETE 101
VACUUM uaocs_threshold_boundary;
VACUUM
SELECT * FROM gp_toolkit.__gp_aovisimap_compaction_info('uaocs_threshold_boundary'::regclass);
NOTICE:  gp_appendonly_compaction_threshold = 10
 content | datafile | compaction_possible | hidden_tupcount | total_tupcount | percent_hidden 
---------+----------+---------------------+-----------------+----------------+----------------
       0 |        2 | f                   |               0 |              0 |           0.00
       1 |        1 | f                   |               0 |              0 |           0.00
       1 |        2 | f                   |               0 |            899 |           0.00
       2 |        2 | f                   |               0 |              0 |           0.00
(4 rows)

